dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,223 @@
+      nvjpeg_memory::AddBuffer<mm::memory_kind::device>(thread_id, device_memory_padding_);
+    }
+  }
+
+  for (int i = 0; i < tp_->NumThreads(); i++) {
+    resources_.emplace_back(nvjpeg_handle_, &device_allocator_, &pinned_allocator_, device_id_);
+  }
+}
+
+NvJpegDecoderInstance::
+PerThreadResources::PerThreadResources(nvjpegHandle_t nvjpeg_handle,
+                                       nvjpegDevAllocator_t *device_allocator,
+                                       nvjpegPinnedAllocator_t *pinned_allocator,
+                                       int device_id)
+  : stream(CUDAStreamPool::instance().Get(device_id)) {
+  CUDA_CALL(nvjpegJpegStreamCreate(nvjpeg_handle, &jpeg_stream));
+  CUDA_CALL(nvjpegBufferPinnedCreate(nvjpeg_handle, pinned_allocator,
+                                     &pinned_buffer));
+  CUDA_CALL(nvjpegBufferDeviceCreate(nvjpeg_handle, device_allocator,
+                                     &device_buffer));
+
+  decode_event = CUDAEvent::Create(device_id);
+
+  auto backend = NVJPEG_BACKEND_HYBRID;  // TODO(msala) allow other backens
+  CUDA_CALL(nvjpegDecoderCreate(nvjpeg_handle, backend, &decoder_data.decoder));
+  CUDA_CALL(nvjpegDecoderStateCreate(nvjpeg_handle, decoder_data.decoder, &decoder_data.state));
+}
+
+NvJpegDecoderInstance::PerThreadResources::PerThreadResources(PerThreadResources&& other)
+: decoder_data(other.decoder_data)
+, device_buffer(other.device_buffer)
+, pinned_buffer(other.pinned_buffer)
+, jpeg_stream(other.jpeg_stream)
+, stream(std::move(other.stream))
+, decode_event(std::move(other.decode_event)) {
+  other.decoder_data = {};
+  other.device_buffer = nullptr;
+  other.pinned_buffer = nullptr;
+  other.jpeg_stream = nullptr;
+}
+
+NvJpegDecoderInstance::~NvJpegDecoderInstance() {
+  DeviceGuard dg(device_id_);
+
+  for (auto &thread_id : tp_->GetThreadIds()) {
+    nvjpeg_memory::DeleteAllBuffers(thread_id);
+  }
+
+  // Call destructors of all thread resources.
+  resources_.clear();
+
+  CUDA_CALL(nvjpegDestroy(nvjpeg_handle_));
+}
+
+NvJpegDecoderInstance::PerThreadResources::~PerThreadResources() {
+  // This check should probably be enough
+  if (stream) {
+    CUDA_CALL(cudaStreamSynchronize(stream));
+  }
+
+  if (jpeg_stream) {
+    CUDA_CALL(nvjpegJpegStreamDestroy(jpeg_stream));
+  }
+  if (pinned_buffer) {
+    CUDA_CALL(nvjpegBufferPinnedDestroy(pinned_buffer));
+  }
+  if (device_buffer) {
+    CUDA_CALL(nvjpegBufferDeviceDestroy(device_buffer));
+  }
+
+  if (decoder_data.decoder && decoder_data.state) {
+    CUDA_CALL(nvjpegDecoderDestroy(decoder_data.decoder));
+    CUDA_CALL(nvjpegJpegStateDestroy(decoder_data.state));
+  }
+}
+
+bool NvJpegDecoderInstance::SetParam(const char *name, const any &value) {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    device_memory_padding_ = any_cast<size_t>(value);
+    return true;
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    host_memory_padding_ = any_cast<size_t>(value);
+    return true;
+  }
```

Comment:
Missing handling of `nvjpeg_num_threads` - also, it would be nice to prevent any changes to this value once tp_ is alive
```suggestion
  } else if (!strcmp(name, "nvjpeg_num_threads") {
    int n = any_cast<int>(value);
    DALI_ENFORCE(n >= 1, make_string("Invalid number of threads specified: ", n));
    if (tp_ && n != tp_->NumThreads())
      DALI_FAIL("Cannot change the number of threads.");
    num_threads_ = n;
    return true;
  }
```